### DISCUSSION TITLE
TextCat component with external embeddings (e.g. Universal Sentence Encoder) possible?

### DISCUSSION THREAD

**Discussion by repodiac** [explosion/spaCy#8848] [Help: Coding & Implementations] [answered]
Hi,

I find it hard to get the information I need from the documentation... in my eyes spaCy 3.x definitely worsened in terms of accessibility, comprehensiveness and "learning by example"... :/

My problem is: I have precomputed **external** embeddings which I manually add as "vector" hook to a Doc object, holding the original text **and** as user_hook the vector representation of the document as given by Google's Universal Sentence Embedding encoding.

I know would like to make use of spaCy's TextCat component for text categorization based **solely** (as input) on the respective embeddings. Is this possible?

The `tok2vec` component as part of TextCat confuses me as if tokenization and/or vectorization is mandatory with TextCat?
Would that mean, my "hack" to introduce USE embeddings does not work?

Thanks for your help in advance!

**Comment by polm** [comment, marked as answer]
You can do this by implementing a custom tok2vec layer, which is a little involved, though assuming you just want to do feature extraction it shouldn't be too bad. 

What you can do is register a custom architecture, maybe call it "DocFeatureExtractor", that you can use to create your own tok2vec. We don't have any guides for this, but you can use the existing [FeatureExtractor][], which is for token-level attributes, as a reference. You can then just specify that as the tok2vec architecture in your config.

[FeatureExtractor]: https://github.com/explosion/spaCy/blob/0a1e299d30356ff58f94116a7bdc7c2c540d44ab/spacy/ml/featureextractor.py#L9

> The tok2vec component as part of TextCat confuses me as if tokenization and/or vectorization is mandatory with TextCat?

A tok2vec takes a list of tokens and converts them to vector representations. So it doesn't do tokenization, that's a separate, earlier step. And textcat does require vector representations, like most modern models. 

> I find it hard to get the information I need from the documentation... in my eyes spaCy 3.x definitely worsened in terms of accessibility, comprehensiveness and "learning by example"... :/

Our intent has been to make spaCy v3 the best version of spaCy ever, can you expand on what's missing? I understand that in this case you couldn't find what you were looking for, but it sounds like you're more broadly unsatisfied.

**Reply by repodiac** [reply]
Hi @polm - sorry for the long delay with responding to your message

> You can do this by implementing a custom tok2vec layer, which is a little involved, though assuming you just want to do feature extraction it shouldn't be too bad.

I would like to do text classification only.
And I was hoping for a rather simple solution in contrast to completely write a custom pipeline replacement for the tok2vec component myself. That seems not to be the case, obviously.
 
> A tok2vec takes a list of tokens and converts them to vector representations. So it doesn't do tokenization, that's a separate, earlier step. And textcat does require vector representations, like most modern models.

I agree. Since Universal Sentence Encoder (USE) is a document embedding method, I can't neither do with tokens or token-based vectors!? I need a method to summarize and feed all this into some compatible format to the `textcat` module. At least that is my understanding so far.

So, do you say that only word2vec like embeddings are possible currently with SpaCy? Or is there a way to feed documents as a whole regardless of words and tokens and thus vectors based on single items? If it's possible, that should work with something like USE as well.

> Our intent has been to make spaCy v3 the best version of spaCy ever, can you expand on what's missing? I understand that in this case you couldn't find what you were looking for, but it sounds like you're more broadly unsatisfied.

I think in general, configuration and settings took over with SpaCy in my eyes. Until 2.x it was possible to simply plug together the base modules and do a little parameter-based configuration in code. Now it seems, I need pages of YAML or JSON config before I can even start anything.

In addition, the examples or tutorials now become overly too complex, I think so that "learning by doing" still worked here (as it did greatly in the past). I really miss the basic usability with SpaCy 2.x which was great and comfortable.

Sorry, if that doesn't make my points clear enough. I certainly cannot express it better at the moment.

**Reply by polm** [reply]
Thanks for the extra details and feedback, that's very helpful. 

Before addressing some of your specific questions, I have a question about your approach to the problem. Why do you specifically want to use USE vectors here? Have you tried using the default spaCy models on your problem? While I have heard that USE is a good model, I'm unclear why you specifically want to use it in this situation. 

Another thing is that if you're using USE and want to do text classification, but not use any other spaCy features, then spaCy doesn't offer very much - classification based on vectors is very simple from an architectural perspective, unlike, say, NER, and can be implemented in any neural framework. 

-----

OK, addressing specific questions...

> I would like to do text classification only.

Not sure if this is a misunderstanding, but when I say "feature extraction" I mean you don't want to update your USE model (which I assume is the case). You could also set it up to be trained in spaCy but that would be more involved.

> And I was hoping for a rather simple solution in contrast to completely write a custom pipeline replacement for the tok2vec component myself. That seems not to be the case, obviously.

This really doesn't have to be very complicated. You can just use a custom architecture for tok2vec, it would look something like this:

```
def get_vecs(doc):
    # this is your custom vec-getting function
    pass

@registry.architectures("custom-vecs.v1")
def build_custom_vectors():
    def forward(docs):
        vecs = [get_vecs(doc) for doc in docs]
        # trivial backprop since we don't learn
        backward = lambda grad: docs
        return vecs, backward

    model = Model("CustomVec", forward=forward)
    return model
```

Admittedly you have to understand a lot to figure this out, but it's not a feature we get many requests for.

> I think in general, configuration and settings took over with SpaCy in my eyes. Until 2.x it was possible to simply plug together the base modules and do a little parameter-based configuration in code. Now it seems, I need pages of YAML or JSON config before I can even start anything.

Thanks for the feedback. We're working on ways to make the configs less verbose, so hopefully that should get better over time.

**Reply by repodiac** [reply]
> Thanks for the extra details and feedback, that's very helpful.

You're welcome. Sorry for the delayed reply, on my side now. And thanks a lot for the detailed answer!

> Before addressing some of your specific questions, I have a question about your approach to the problem. Why do you specifically want to use USE vectors here? Have you tried using the default spaCy models on your problem? While I have heard that USE is a good model, I'm unclear why you specifically want to use it in this situation.

The reason why I prefer USE over "pure SpaCy" is that USE has a very good document embedding. SpaCy uses some external word embeddings or even word embeddings based on the dataset solely (for supervised tasks in general), right? I just want to replace the SpaCy embeddings with external USE embeddings and then use the optimized (both in runtime and quality) architecture of SpaCy for text classification. That's all.
 
> Another thing is that if you're using USE and want to do text classification, but not use any other spaCy features, then spaCy doesn't offer very much - classification based on vectors is very simple from an architectural perspective, unlike, say, NER, and can be implemented in any neural framework.

That's true, but SpaCy is blazing fast on CPU... and I don't have to fiddle with good defaults, in addition. Just my experience...

> OK, addressing specific questions...
> 
> > I would like to do text classification only.
> 
> Not sure if this is a misunderstanding, but when I say "feature extraction" I mean you don't want to update your USE model (which I assume is the case). You could also set it up to be trained in spaCy but that would be more involved.

No, no gradient backprop to the embeddings, that's correct.

> > And I was hoping for a rather simple solution in contrast to completely write a custom pipeline replacement for the tok2vec component myself. That seems not to be the case, obviously.
> 
> This really doesn't have to be very complicated. You can just use a custom architecture for tok2vec, it would look something like this:
> 
> ```
> def get_vecs(doc):
>     # this is your custom vec-getting function
>     pass
> 
> @registry.architectures("custom-vecs.v1")
> def build_custom_vectors():
>     def forward(docs):
>         vecs = [get_vecs(doc) for doc in docs]
>         # trivial backprop since we don't learn
>         backward = lambda grad: docs
>         return vecs, backward
> 
>     model = Model("CustomVec", forward=forward)
>     return model
> ```
> 
> Admittedly you have to understand a lot to figure this out, but it's not a feature we get many requests for.

I see. But: GREAT! Looks really simple. I will try and check it out, thanks!

> > I think in general, configuration and settings took over with SpaCy in my eyes. Until 2.x it was possible to simply plug together the base modules and do a little parameter-based configuration in code. Now it seems, I need pages of YAML or JSON config before I can even start anything.
> 
> Thanks for the feedback. We're working on ways to make the configs less verbose, so hopefully that should get better over time.

+1

**Reply by polm** [reply]
> The reason why I prefer USE over "pure SpaCy" is that USE has a very good document embedding. SpaCy uses some external word embeddings or even word embeddings based on the dataset solely (for supervised tasks in general), right? 

No, that is not an accurate description of how the spaCy embeddings work. 

The pretrained spaCy pipelines sometimes include word vectors, which are sourced externally. Those word vectors are used as input features in training. Training results in task-specific models (parser, tagger, NER), and also a shared tok2vec or Transformer layer. Leaving aside the Transformers, the tok2vec is a 1D CNN, so while it's not as powerful as USE or Transformers it's still contextual embeddings. The spaCy models are trained on large annotated corpora, and while the details vary by language it's usually formal text like newspapers. 

None of the spaCy models just use word vectors. If you train your own model, there are different ways you can configure it, but it would be pretty common to re-use the base tok2vec and train a task head from scratch. 

Looking at the USE architecture, it seems it uses self-attention like Transformers, but with a specific notion of a document embedding.

If you haven't evaluated the base spaCy models (including Transformers), I would strongly suggest you do so - even if USE is better for your application, I wouldn't expect it to be a night-and-day difference.

**Comment by polm** [comment]
Short update to this - I hadn't noticed it before but there's actually a USE specific package [on the Universe](https://spacy.io/universe/project/spacy-universal-sentence-encoder), and it works with spaCy 3, so you should be able to just use that.

**Reply by repodiac** [reply]
Thanks, I've checked that but it does not make use of textcat or at least it is not clear how the textcat module could use it upfront. It just adds the document vector to the document metadata so that `doc.similarity` could use it for similarity matchings.
